org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/Annotation.java

```diff
@@ -768,37 +769,39 @@ public boolean isRuntimeInvisible() {
 		return (metaTagBits & TagBits.AnnotationRetentionMASK) == TagBits.AnnotationClassRetention;
 	}
 
-	public boolean isRuntimeTypeInvisible() {
+	public boolean isRuntimeTypeInvisible(int targetType) {
 		final TypeBinding annotationBinding = this.resolvedType;
 		if (annotationBinding == null) {
 			return false;
 		}
 		long metaTagBits = annotationBinding.getAnnotationTagBits(); // could be forward reference
 
-		if ((metaTagBits & (TagBits.AnnotationTargetMASK)) != 0) {
-			if ((metaTagBits & (TagBits.AnnotationForTypeParameter | TagBits.AnnotationForTypeUse)) == 0) {
+		if ((metaTagBits & (TagBits.AnnotationTargetMASK)) == 0) { // In the absence of explicit target, applicable only to declaration sites
+			if (targetType != AnnotationTargetTypeConstants.CLASS_TYPE_PARAMETER && targetType != AnnotationTargetTypeConstants.METHOD_TYPE_PARAMETER)
```

Comment:
In my naive understanding, methods `isRuntimeTypeVisible` and `isRuntimeTypeInvisible` would only cover codegen for **type annotations** (declaration scenarii being covered by `isRuntimeVisible`/`isRuntimeInvisible`).
If that were true then parameter `targetType` would be unnecessary, as in absence of a `@Target`  we can directly answer `false`, can't we?

The only test that disagrees is the new `testGH1096()`, challenging a no-target annotation on a type parameter.

Does this mean, annotations on type parameters are handled through the TYPE_USE-related code, rather than like "normal" declaration annotations?

Should we schedule a follow-up issue, to reconsider this?

Reply:
> In my naive understanding, methods `isRuntimeTypeVisible` and `isRuntimeTypeInvisible` would only cover codegen for **type annotations** (declaration scenarii being covered by `isRuntimeVisible`/`isRuntimeInvisible`). If that were true then parameter `targetType` would be unnecessary, as in absence of a `@Target` we can directly answer `false`, can't we?
> 

Consider this program:

```
import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

@Target(ElementType.TYPE_USE)
@interface TA {
}

@TA
public class X {
} 
```

which reaches 

```
SingleMemberAnnotation(Annotation).isRuntimeTypeInvisible(int) line: 777	
ClassFile.generateRuntimeAnnotations(Annotation[], long) line: 4581	
ClassFile.addAttributes() line: 423	
TypeDeclaration.generateCode(ClassFile) line: 772	
TypeDeclaration.generateCode(CompilationUnitScope) line: 831	
```

and org.eclipse.jdt.core.tests.compiler.regression.TypeAnnotationTest.testGH1096()

which reaches the call stack:

```
MarkerAnnotation(Annotation).isRuntimeTypeInvisible(int) line: 777	
TypeReference$AnnotationCollector.internalVisit(Annotation) line: 197	
TypeReference$AnnotationCollector.visit(MarkerAnnotation, BlockScope) line: 244	
MarkerAnnotation.traverse(ASTVisitor, BlockScope) line: 43	
TypeParameter.getAllAnnotationContexts(int, int, List<AnnotationContext>) line: 72	
ClassFile.completeMethodInfo(MethodBinding, int, int) line: 2455	
MethodDeclaration(AbstractMethodDeclaration).generateCode(ClassFile) line: 377	
MethodDeclaration(AbstractMethodDeclaration).generateCode(ClassScope, ClassFile) line: 292	
TypeDeclaration.generateCode(ClassFile) line: 761	
```
In both scenarios the method is supposed to return true.

What is a suitable value for the parameter targetType of the method
org.eclipse.jdt.internal.compiler.ast.Annotation.isRuntimeTypeInvisible(int) for the first call (stack) ??

(Despite it not being relevant - the parameter gets "touched" only when no target is specified) I simply invented a value that is outside of all possible type annotation target locations enumerated by table 4.7.20 of the JVMS. 

There are only three meaningful values that influence the method's behavior:
targetType == AnnotationTargetTypeConstants.CLASS_TYPE_PARAMETER
targetType == AnnotationTargetTypeConstants.METHOD_TYPE_PARAMETER and
targetType being *anything else* - For that -1  given the symbolic name NonJVMS4_7_20_TargetLocation was my choice.

Better ideas welcome. 


> The only test that disagrees is the new `testGH1096()`, challenging a no-target annotation on a type parameter.
> 
> Does this mean, annotations on type parameters are handled through the TYPE_USE-related code, rather than like "normal" declaration annotations?
> 
> Should we schedule a follow-up issue, to reconsider this?